### PR TITLE
Make service dns domain configurable

### PR DESCRIPTION
**Pull request by kfox1111**
Some clusters do not use cluster.local for the root of their service dns domain. This patch makes the chart usable on those clusters.

fixes: https://github.com/spiffe/helm-charts/issues/158

### CODE REVIEW

File: charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml [hunk 1]
```diff
@@ -113,6 +113,7 @@ tolerations: []
 affinity: {}
 
 trustDomain: example.org
+serviceDnsDomain: cluster.local
```
Comment (marcofranssen) [inline charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml:116]: DnsDomain sounds a bit like duplicate naming.

Domain Name Service Domain 🤔 

Maybe just `clusterDomain`?

Comment (marcofranssen) [reply charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml:116]: Fair point.

Maybe add a documentation comment with that explanation so it ends up in the docs.

**Comment by faisal-memon** [comment]
In controller manager we auto detect this by querying `kubernetes.default.svc` and extracting the domain from the reply. Can we do something similar here? https://github.com/spiffe/spire-controller-manager/blob/main/main.go#L336-L348

File: charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml (hunk 1 above)
Comment (kfox1111) [reply charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml:116]: Ok, I can get behind that. I didn't see anything in the helm chart but the spire-controller-manager's config having an option for it makes sense. I'll fix it up.

**Comment by kfox1111** [comment]
> In controller manager we auto detect this by querying `kubernetes.default.svc` and extracting the domain from the reply. Can we do something similar here? https://github.com/spiffe/spire-controller-manager/blob/main/main.go#L336-L348

Interesting idea....

Yes and no.... the oidc configmap has it hardcoded into it by the look of it. So, the only way I can think of to fix that would be to:
1. add an emptydir
2. add an init container:
    1. Do the lookup
    2. cut it down to the subdomain
    3. template out the config from configmap a second time, injecting the discovered domain into it
    4. write the config to the emptydir
3. Modify the main container to read from the emptydir copy of the config instead of the configmap

We may or may not need a custom image to do the lookup too. not sure busybox has enough tooling in it to do step 2.

Seems possible, but is an awful lot of work for something most people don't set on their clusters?

File: charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml (hunk 1 above)
Comment (faisal-memon) [reply charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml]: We havent added it to the helm chart yet. Not sure its been released yet, it was just merged a couple weeks ago.

**Comment by faisal-memon** [comment]
> > In controller manager we auto detect this by querying `kubernetes.default.svc` and extracting the domain from the reply. Can we do something similar here? https://github.com/spiffe/spire-controller-manager/blob/main/main.go#L336-L348
> 
> Interesting idea....
> 
> Yes and no.... the oidc configmap has it hardcoded into it by the look of it. So, the only way I can think of to fix that would be to:
> 
> 1. add an emptydir
> 2. add an init container:
>    
>    1. Do the lookup
>    2. cut it down to the subdomain
>    3. template out the config from configmap a second time, injecting the discovered domain into it
>    4. write the config to the emptydir
> 3. Modify the main container to read from the emptydir copy of the config instead of the configmap
> 
> We may or may not need a custom image to do the lookup too. not sure busybox has enough tooling in it to do step 2.
> 
> Seems possible, but is an awful lot of work for something most people don't set on their clusters?

Does seem like a lot of work. I was thinking a pre install hook. We can always add an auto detect feature later if we think of a good way to do it.

**Comment by kfox1111** [comment]
> 
> Does seem like a lot of work. I was thinking a pre install hook. We can always add an auto detect feature later if we think of a good way to do it.

In general, I dislike hooks. and prefer if they can be avoided. I'd prefer the initcontainer thing over a hook.
I'm also slightly skeptical that the cluster domain can always be reliably detected. So a manual override may still be needed anyway?

Yeah, maybe we merge this and then consider a separate pr?

**Comment by kfox1111** [comment]
Oh. I think I may understand the question. Did you mean:
Should we be setting clusterDomain: cluster.local in charts/spire/values.yaml ?

**Comment by faisal-memon** [comment]
> Oh. I think I may understand the question. Did you mean: Should we be setting clusterDomain: cluster.local in charts/spire/values.yaml ?

Yes, thats what i meant

**Comment by kfox1111** [comment]
Yeah. I can add that.

**Review by marcofranssen — CHANGES_REQUESTED** [review]
.

File: charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml [hunk 2]
```diff
@@ -113,6 +113,8 @@ tolerations: []
 affinity: {}
 
 trustDomain: example.org
+# -- Update this if you used the kubeadm init --service-dns-domain flag
```
Comment (marcofranssen) [inline charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml:116]:
```suggestion
# -- This is the value of your clusters `kubeadm init --service-dns-domain` flag
```

File: charts/spire/values.yaml [hunk 3]
```diff
@@ -10,6 +10,8 @@ fullnameOverride: ""
 # subcharts
 spire-server:
   enabled: true
+  # -- Update this if you used the kubeadm init --service-dns-domain flag
```
Comment (marcofranssen) [inline charts/spire/values.yaml:13]:
```suggestion
# -- This is the value of your clusters `kubeadm init --service-dns-domain` flag
```

File: charts/spire/values.yaml [hunk 4]
```diff
@@ -21,6 +23,8 @@ spire-server:
 
 spire-agent:
   enabled: true
+  # -- Update this if you used the kubeadm init --service-dns-domain flag
```
Comment (marcofranssen) [inline charts/spire/values.yaml:26]:
```suggestion
# -- This is the value of your clusters `kubeadm init --service-dns-domain` flag
```

File: charts/spire/charts/spire-server/values.yaml [hunk 5]
```diff
@@ -86,6 +86,8 @@ clusterName: example-cluster
 trustDomain: example.org
 
 bundleConfigMap: spire-server
+# -- Update this if you used the kubeadm init --service-dns-domain flag
```
Comment (marcofranssen) [inline charts/spire/charts/spire-server/values.yaml:89]:
```suggestion
# -- This is the value of your clusters `kubeadm init --service-dns-domain` flag
```

**Comment by marcofranssen** [comment]
If we can discuss and settle on #156 approach we can apply same pattern here on these changes.

File: charts/spire/values.yaml [hunk 6]
```diff
@@ -7,6 +7,8 @@ global:
     trustDomain: ""
     # -- Override all instances of bundleConfigMap
     bundleConfigMap: ""
+    # -- This is the value of your clusters `kubeadm init --service-dns-domain` flag
+    clusterDomain: ""
```
Comment (marcofranssen) [inline charts/spire/values.yaml:11]: This is a non spire var.

Lift one level up.

    global.clusterDomain

**Review by marcofranssen — CHANGES_REQUESTED** [review]

File: charts/spire/values.yaml (hunk 6 above)
Comment (kfox1111) [reply charts/spire/values.yaml]: And clusterDomain, while makes sense in the context of spire, could mean different things in the context of a chart depending on the spire chart. Its not very specific a value.

File: charts/spire/values.yaml [hunk 7]
```diff
@@ -1,6 +1,8 @@
 # You can enable config/features that affect all services here.
 global:
   spire:
+    # -- Set to the value of your clusters `kubeadm init --service-dns-domain` flag
+    clusterDomain: "cluster.local"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:5]:
```suggestion
global:
  # -- Set to the value of your clusters `kubeadm init --service-dns-domain` flag
  clusterDomain: "cluster.local"

  spire:
```

This isn't a spire setting. Think its better of at root level.

File: charts/spire/values.yaml [hunk 8]
```diff
@@ -1,5 +1,8 @@
 # You can enable config/features that affect all services here.
 global:
+  k8s:
+    # -- Set to the value of your clusters `kubeadm init --service-dns-domain` flag
```
Comment (marcofranssen) [inline charts/spire/values.yaml:4]:
```suggestion
# -- This is the value of your clusters `kubeadm init --service-dns-domain` flag
```

To align with the other descriptions, also requires doc update.

File: charts/spire/values.yaml [hunk 9]
```diff
@@ -1,5 +1,8 @@
 # You can enable config/features that affect all services here.
 global:
+  k8s:
+# -- This is the value of your clusters `kubeadm init --service-dns-domain` flag
+    clusterDomain: "cluster.local"
```
Comment (marcofranssen) [inline charts/spire/values.yaml:5]:
```suggestion
    # -- This is the value of your clusters `kubeadm init --service-dns-domain` flag
    clusterDomain: "cluster.local"
```

**Review by marcofranssen — APPROVED** [review]
LGTM 🚀

**Pull request merged**